### PR TITLE
Mdd m borg/changes for dumping in silos

### PR DESCRIPTION
**Pull request by MddMBorg**
I had noticed 2 things while playing:

When using a worker to push chaff through a bunker, the worker would often push the chaff too far and out the other end. I made a change in bunkersiloaidriver to push the chaff a little bit less to lessen how much chaff will get pushed out of an open-ended bunker.

When using reverse-unloading (such as into a silo), the unloader would slowly creep forward after consecutive unloads, since the heap would slowly creep forward, even with a worker using a shield to push it through (getting chaff too fast from chopper). I had made a change in AIDriver that checks the size of the heap before deciding to unload. My geometry is probably off, but it gets the fill level of a 1 m^2 area at the discharge node of the tipper. If the fill level is < 2000, the tipper will keep reversing (over the heap), until the fill level is high enough. This means that the tipper can go further back onto the heap, and dumps the chaff/product onto an existing pile, allowing it to pile up higher, instead of stopping as soon as it reaches even a small patch of chaff on the bunker floor that has been missed by the worker.

I'd be happy to take feedback on any possible better approaches. For instance, I just picked the 2000 number because it seemed that it fit, but it's possible that smaller trailers/tractors would need a smaller number (the heap may be too big to reverse onto).

### CODE REVIEW

**Comment by Tensuko** [comment]
> When using a worker to push chaff through a bunker, the worker would often push the chaff too far and out the other end. I made a change in bunkersiloaidriver to push the chaff a little bit less to lessen how much chaff will get pushed out of an open-ended bunker.

Screenshots of the Issue ?
The BunkersiloAIDriver got a lot of changes with 6.4.0.X and should never drive to far to the end on those Bunkers.
We did test that with GIANTS Bunkers and it was working just fine.


> I'd be happy to take feedback on any possible better approaches. For instance, I just picked the 2000 number because it seemed that it fit, but it's possible that smaller trailers/tractors would need a smaller number (the heap may be too big to reverse onto).

We have to make sure, that all kind of trailers and tractors will have no problem, to drive up the heap.
So it should be only a very little amount.
Also here I would like to see a screenshot how that looks like.
If to high, it have to be decreased to not get smaler vehicles stuck.

Beside that, thanks for your contribution! Those tweaks are very welcome, but pls consider more then just your current situation for such changes :)

**Comment by MddMBorg** [comment]
ChaffTooFar

Above is an instance of the chaff being pushed out of the bunker by the silo worker. This is the base silo that can be placed from the store. The furthermost parts are where it was being pushed to originally, before my change. The chaff is still sometimes pushed a bit too far, but less so (hard to demonstrate in a picture).

Dumping1
Approx position of where dumping would occur before changes.

Dumping2
This is where the changed AIDriver got to (Mode 2) before unloading( which has almost all 3 wheels of the tipper backed up over the heap.

Dumping3
The chaff level isn't very high there, maybe 1-1.5 meters high. The biggest issue would be a pile that is very steep instead of a shallow slope like in the picture. I could have it change the cutoff based on the horsepower of the tractor, or capacity of the tipper?

Lemme know if you want more pics.

**Comment by Tensuko** [comment]
> The chaff is still sometimes pushed a bit too far, but less so (hard to demonstrate in a picture).

Well, that is okay then, would be to much if we drive only 3/4 of the Bunker :)

Pics after maybe gsSetDayTime 10 for more brightness :)

> I could have it change the cutoff based on the horsepower of the tractor, or capacity of the tipper?

Maybe depending on the capacity of the trailer could make sence.
I might play arround with the numbers for some adjustment later.

But basically I rly like that :) Thanks!

**Comment by Tensuko** [comment]
Works fine like that, had no Issues with smal Trailers so far, unloading on different heaps.

File: AIDriver.lua [hunk 1]
```diff
@@ -1280,6 +1282,15 @@ function AIDriver:checkForHeapBehindMe(tipper)
 	end
 end
 
+function AIDriver:getHeapSizeBehindMe(tipper)
```
Comment (schwiti6190) [inline AIDriver.lua:1285]: @MddMBorg Hi there,

A small code improvement would be to consolidate the functions: ` self:checkForHeapBehindMe(tipper), self:getHeapSizeBehindMe(tipper)`into one like this: 
```
`function AIDriver:checkForHeapBehindMe(tipper)
    local dischargeNode = tipper:getCurrentDischargeNode().node
    local offset = -self.settings.loadUnloadOffsetZ:get()
    offset = courseplay:isNodeTurnedWrongWay(self.vehicle,dischargeNode)and -offset or offset
    local startX,startY,startZ = localToWorld(dischargeNode,0,0,offset) ;
    local tempHeightX,tempHeightY,tempHeightZ = localToWorld(dischargeNode,0,0,offset+0.5) 
    local searchWidth = 1    
    local fillType = DensityMapHeightUtil.getFillTypeAtLine(startX,startY,startZ,tempHeightX,tempHeightY,tempHeightZ, searchWidth)
    if fillType == tipper.cp.fillType then
        local fillLevel = DensityMapHeightUtil.getFillLevelAtArea(fillType,startX,startZ,startX,startZ+1,startX+1,startZ)    --should represent an area of 1m^2 starting at the discharge node
        return true,fillLevel
    end
    return false,0
end`
```

File: AIDriver.lua [hunk 2]
```diff
@@ -1193,9 +1193,11 @@ function AIDriver:dischargeAtUnloadPoint(dt,unloadPointIx)
 				z = courseplay:isNodeTurnedWrongWay(vehicle,tipRefpoint)and -z or z
 
 				local foundHeap = self:checkForHeapBehindMe(tipper)
```
Comment (schwiti6190) [inline AIDriver.lua:1195]: This one would be: `local foundHeap,fillLevel = self:checkForHeapBehindMe(tipper)`

**Review by schwiti6190 — CHANGES_REQUESTED** [review]

File: AIDriver.lua (hunk 2 above)
Comment (MddMBorg) [reply AIDriver.lua]: I've made those changes in latest commit

Comment (schwiti6190) [reply AIDriver.lua]: @MddMBorg Nice work. Thank you :)

**Pull request merged**